packages/react/src/utils/useIOSKeyboardSlideFix.ts

```diff
@@ -0,0 +1,78 @@
+import * as React from 'react';
+import type { ElementProps } from '@floating-ui/react';
+import { getTarget, isTypeableElement } from '@floating-ui/react/utils';
+import { useEnhancedEffect } from './useEnhancedEffect';
+import { isIOS } from './detectBrowser';
+import { ownerDocument } from './owner';
+import { useEventCallback } from './useEventCallback';
+import { getPreventScrollCount } from './useScrollLock';
+
+/**
+ * Temporarily allows disabling the scroll lock when a typeable element (e.g. input or textarea)
+ * is focused. This allows the document to slide up when the keyboard is shown on iOS.
+ * Scroll needs to be manually restored since the sliding causes the page to scroll in the
+ * background.
+ * https://github.com/mui/base-ui/issues/1455
+ */
+export function useIOSKeyboardSlideFix(params: {
+  enabled: boolean;
+  popupRef: React.RefObject<HTMLElement | null>;
+  setLock: (lock: boolean) => void;
+}) {
+  const { enabled, setLock: setLockParam, popupRef } = params;
+
+  const setLock = useEventCallback(setLockParam);
+
+  const scrollRef = React.useRef({ x: 0, y: 0 });
+  const hasBeenEnabledRef = React.useRef(enabled);
+
+  useEnhancedEffect(() => {
+    if (!isIOS()) {
+      return undefined;
+    }
+
+    if (enabled) {
+      hasBeenEnabledRef.current = true;
+    }
+
+    if (!hasBeenEnabledRef.current) {
+      return undefined;
+    }
+
+    const doc = ownerDocument(popupRef.current);
+    const html = doc.documentElement;
+    if (enabled) {
+      scrollRef.current = { x: html.scrollLeft, y: html.scrollTop };
+      return () => {
+        if (getPreventScrollCount() === 0) {
+          html.scrollTop = scrollRef.current.y;
+          html.scrollLeft = scrollRef.current.x;
+        }
+      };
+    }
+
+    return undefined;
+  }, [enabled, popupRef]);
+
+  const handlePress = useEventCallback((event: React.FocusEvent | React.MouseEvent) => {
+    if (!isIOS()) {
+      return;
+    }
+
+    const target = getTarget(event.nativeEvent) as Element | null;
+    if (isTypeableElement(target)) {
+      setLock(false);
+      setTimeout(() => setLock(true));
+    }
+  });
+
+  const floating: ElementProps['floating'] = React.useMemo(
+    () => ({
+      onFocus: handlePress,
+      onClick: handlePress,
+    }),
+    [handlePress],
+  );
+
+  return React.useMemo(() => ({ floating }), [floating]);
```

Comment:
Let's merge these two `useMemo`s

Reply:
Each one needs to be memo'd separately

Reply:
Why can't we make it
```
return React.useMemo(
  () => ({
    floating: {
      onFocus: handlePress,
      onClick: handlePress,
    },
  }),
  [handlePress],
);
```

Reply:
`useInteractions` uses `.floating` property as a dependency in `useCallback`

Reply:
Does it matter? The value returned from this hook is the same regardless of how many memos are used, no? It's just a matter of optimization.
The memoized value won't ever change anyway as `handlePress` is stable.

Reply:
Yeah you're right - the nested object should also be stable in this case since the top level object itself is stable between renders